File: problem3.js
// Problem 3 from Project Euler
// The prime factors of 13195 are 5, 7, 13 and 29.
//
// What is the largest prime factor of the number 600851475143?
//
// Solution: 6857

var input = 600851475143;

var isPrime = function(num) {
    for (var i = 2; i < num; i++) {
        if (num % i === 0) {
            return false;
        }
    }
    return true;
}

var max = 0;
var factors = [];

// do not need to go farther than sqrt(n), if n is divisible by p,
// then n = p x q and if q were smaller than p, q would have already
// been detected.
for (var i = 0; i < Math.pow(input, 0.5); i++) {
    if (input % i === 0) {
        if (isPrime(i)) {
            factors.push(i);
            if (i > max) {
                max = i;
            }
        }
    }
}

console.log("Solution:", 6857);
console.log(max);
console.log(factors);

File: problem3.rb
# Problem 3 from Project Euler
# The prime factors of 13195 are 5, 7, 13 and 29.
#
# What is the largest prime factor of the number 600851475143?
#
# Solution: 6857

input = 600851475143 

def isPrime(n)
  (2...n).each {|i| return false if n % i == 0}
  return true
end

factors = []
# do not need to go farther than sqrt(n), if n is divisible by p,
# then n = p x q and if q were smaller than p, q would have already
(2...Math.sqrt(input)).each do |i|
# been detected.
  # check if input is divisible by i
  # if not, no point in checking if i is prime
  factors << i if (input % i == 0) and (isPrime(i))
end

puts "Solution:", 6857
puts factors[-1]
